compiler/rustc_mir/src/monomorphize/polymorphize.rs

```diff
@@ -27,21 +27,25 @@ pub fn provide(providers: &mut Providers) {
     providers.unused_generic_params = unused_generic_params;
 }
 
-/// Determine which generic parameters are used by the function/method/closure represented by
-/// `def_id`. Returns a bitset where bits representing unused parameters are set (`is_empty`
+/// Determine which generic parameters are used by the `instance`.
+///
+/// Returns a bitset where bits representing unused parameters are set (`is_empty`
 /// indicates all parameters are used).
-fn unused_generic_params(tcx: TyCtxt<'_>, def_id: DefId) -> FiniteBitSet<u32> {
-    debug!("unused_generic_params({:?})", def_id);
+fn unused_generic_params<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    instance: InstanceDef<'tcx>,
+) -> FiniteBitSet<u32> {
+    debug!("unused_generic_params({:?})", instance);
 
+    // If polymorphization disabled, then all parameters are used.
     if !tcx.sess.opts.debugging_opts.polymorphize {
-        // If polymorphization disabled, then all parameters are used.
         return FiniteBitSet::new_empty();
     }
 
-    // Polymorphization results are stored in cross-crate metadata only when there are unused
-    // parameters, so assume that non-local items must have only used parameters (else this query
-    // would not be invoked, and the cross-crate metadata used instead).
-    if !def_id.is_local() {
+    // Exit early if this instance should not be polymorphized.
+    let def_id = instance.def_id();
```

Comment:
Before rebasing, `should_polymorphize` invoked `def_id_if_not_guaranteed_local_codegen` instead of using this value. It returned `true` whenever `def_id_if_not_guaranteed_local_codegen` returned `None`, which meant that a `ClosureOnceShim` (particularly, the shim for `FnOnce::fn_once`) wasn't being checked for available MIR, which was causing an ICE after I rebased.